### ISSUE TITLE
grep: memory exhausted

### ISSUE THREAD

**Issue by anarchy9388** [anarchy9388/scan#9]

**Comment by anarchy9388** [comment]
lglf132:/tmp/mark # python ./scan.py scan -f ./chunkids -s /tmp/free_blocks_with_history_83317d0e-3152-442a-a9cd-ec8424710aff_2019-07-08-14-36-19
scan.py v0.2.0                 Started 2019-07-08 21:28:32

Log file will be saved in: /tmp/scan_busy_blocks.log.2019-07-08-21-28-32
Result will be saved in dir: /tmp/block_scan_res
Total: 12362 candidate blocks need to pre-check.

grep: memory exhausted
grep: memory exhausted

**Comment by anarchy9388** [comment]
@MarkHe1222 

Updated to use grep -oPaz:

The grep program reads a line at a time into memory. A line is defined as everything after one newline character and up to the next one. With binary data, there could be a very large space without any newlines.

You could try using grep -z. This tells grep to treat null bytes as the input record separator instead of newlines. Extremely large chunks of binary data are less likely to contain no null byte than no newline. In practice, the most likely chunk of data with no newline is a long sequence of null bytes in an area of the disk that's never been written to yet. Large amounts of text data containing no null bytes are likely to not be so large as to exhaust memory. Another benefit of grep -z is that the output will contain whole blocks (typically 1–4kB) of text, not just one line.

**Comment by anarchy9388** [comment]
still hitting Memory Exhausted with -z:
`scan.py v0.2.0                 Started 2019-07-08 21:47:37

Log file will be saved in: /tmp/scan_busy_blocks.log.2019-07-08-21-47-37
Result will be saved in dir: /tmp/block_scan_res
Total: 12362 candidate blocks need to pre-check.

grep: memory exhausted
`

**Comment by anarchy9388** [comment]
@MarkHe1222 

Reducing the threads to 2 here with the -z seems to work the best.

**Comment by MarkHe1222** [comment]
> lglf132:/tmp/mark # python ./scan.py scan -f ./chunkids -s /tmp/free_blocks_with_history_83317d0e-3152-442a-a9cd-ec8424710aff_2019-07-08-14-36-19
> scan.py v0.2.0 Started 2019-07-08 21:28:32
> 
> Log file will be saved in: /tmp/scan_busy_blocks.log.2019-07-08-21-28-32
> Result will be saved in dir: /tmp/block_scan_res
> Total: 12362 candidate blocks need to pre-check.
> 
> grep: memory exhausted
> grep: memory exhausted

Hi @anarchy9388  is the problem still existed after you add the option "-z" ?

**Comment by anarchy9388** [comment]
@MarkHe1222 

Issue still exists even with just 2 threads and using grep.

I started writing a different way, this should be much better process the file in chunks and use binary regex. more accurate and grep wont hit memory exhausted issue.

```
import os
import re
import sys
from multiprocessing import Pool

pool = Pool(processes=4)
block_file = '/dae/uuid-b8d496c8-e542-4d03-9860-f0109bf4dfb0/0045'
chunk_size = 8388608
block_file_size = os.path.getsize(block_file)
file_chunks = range(chunk_size, (block_file_size + chunk_size), chunk_size)
chunks = []
print chunk_size
print block_file_size
re_chunks = re.compile(b'[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}').findall

def process_file_chunk(f, position):
    f.seek(position, 0)
    file_chunk = f.read(chunk_size)
    tmp_chunks = re_chunks(file_chunk)
    return tmp_chunks

procs = []
for position in range(chunk_size, (block_file_size + chunk_size), chunk_size):
    file_handle = open(block_file, 'rb')
    proc = pool.apply_async(process_file_chunk, (file_handle, position, ))
    procs.append(proc)
    proc.get()

for proc in procs:
    while not proc.ready():
        proc.wait(1)

for proc in procs:
    print proc.get()




pool.close()
pool.join()
```

**Comment by MarkHe1222** [comment]
Thanks @anarchy9388 , I will take a look today.

**Comment by anarchy9388** [comment]
@MarkHe1222  No need, verified this is fixed using new method in branch 0.3 :)
